File: api/src/services/suggestionsService.ts
import { createHash } from "node:crypto";
import {
  SWAP_SUGGESTIONS_COUNT,
  type SwapSuggestionsResponse,
  type SwapSuggestion,
  type SwapSuggestionsRequest,
} from "@carboniq/contracts";
import {
  CATEGORY_ORDER,
  CATEGORY_ORDER_INDEX,
  SUGGESTION_RULES,
  type EmissionCategoryValue,
} from "../lib/aiRules.js";
import { roundTo } from "../lib/aiMath.js";
import { emissionsService } from "./emissionsService.js";

type SuggestionNarrativeInput = {
  currentCategory: EmissionCategoryValue;
  currentDescription: string;
  alternativeDescription: string;
  currentCo2eMonthly: number;
  alternativeCo2eMonthly: number;
  co2eSavingsMonthly: number;
  priceDifferenceUsd: number;
  difficulty: SwapSuggestion["difficulty"];
};

type SuggestionNarrativeText = {
  currentDescription: string;
  alternativeDescription: string;
};

class TemplateSuggestionNarrator {
  narrate(suggestion: SuggestionNarrativeInput): SuggestionNarrativeText {
    return {
      currentDescription: suggestion.currentDescription,
      alternativeDescription: suggestion.alternativeDescription,
    };
  }
}

class SuggestionsService {
  private readonly narrator = new TemplateSuggestionNarrator();
  private readonly narrationCache = new Map<string, SuggestionNarrativeText>();

  private rankCategories(
    categoryTotals: Record<EmissionCategoryValue, number>,
    requestedCategories?: EmissionCategoryValue[]
  ): EmissionCategoryValue[] {
    const requested = requestedCategories?.length
      ? requestedCategories
      : CATEGORY_ORDER;
    const ranked = [...requested].sort(
      (left, right) =>
        categoryTotals[right] - categoryTotals[left] ||
        CATEGORY_ORDER_INDEX[left] - CATEGORY_ORDER_INDEX[right]
    );

    if (ranked.length >= SWAP_SUGGESTIONS_COUNT.min) {
      return ranked.slice(0, SWAP_SUGGESTIONS_COUNT.max);
    }

    const seen = new Set(ranked);
    const backfill = [...CATEGORY_ORDER]
      .sort(
        (left, right) =>
          categoryTotals[right] - categoryTotals[left] ||
          CATEGORY_ORDER_INDEX[left] - CATEGORY_ORDER_INDEX[right]
      )
      .filter((category) => !seen.has(category));

    return [...ranked, ...backfill].slice(0, SWAP_SUGGESTIONS_COUNT.max);
  }

  private narrate(
    wallet: string,
    suggestion: SuggestionNarrativeInput
  ): SuggestionNarrativeText {
    const cacheKey = createHash("sha256")
      .update(
        [
          wallet,
          suggestion.currentCategory,
          suggestion.currentCo2eMonthly,
          suggestion.alternativeCo2eMonthly,
          suggestion.co2eSavingsMonthly,
          suggestion.priceDifferenceUsd,
          suggestion.difficulty,
        ].join("|")
      )
      .digest("hex");
    const cached = this.narrationCache.get(cacheKey);
    if (cached) {
      return cached;
    }

    const narration = this.narrator.narrate(suggestion);
    this.narrationCache.set(cacheKey, narration);
    return narration;
  }

  getSwapSuggestions(
    request: SwapSuggestionsRequest
  ): SwapSuggestionsResponse {
    const snapshot = emissionsService.getCanonicalSnapshot(request.wallet);
    const rankedCategories = this.rankCategories(
      snapshot.categoryEmissionTotals,
      request.categories
    );

    const suggestions: SwapSuggestion[] = [];
    let totalPotentialSavingsMonthly = 0;

    for (const category of rankedCategories) {
      const rule = SUGGESTION_RULES[category];
      const currentCo2eMonthly = snapshot.categoryEmissionTotals[category];
      const alternativeCo2eMonthly = roundTo(
        currentCo2eMonthly * rule.alternativeShare,
        2
      );
      const co2eSavingsMonthly = roundTo(
        currentCo2eMonthly - alternativeCo2eMonthly,
        2
      );
      const narration = this.narrate(request.wallet, {
        currentCategory: category,
        currentDescription: rule.currentDescription,
        alternativeDescription: rule.alternativeDescription,
        currentCo2eMonthly,
        alternativeCo2eMonthly,
        co2eSavingsMonthly,
        priceDifferenceUsd: rule.priceDifferenceUsd,
        difficulty: rule.difficulty,
      });

      suggestions.push({
        currentCategory: category,
        currentDescription: narration.currentDescription,
        currentCo2eMonthly,
        alternativeDescription: narration.alternativeDescription,
        alternativeCo2eMonthly,
        co2eSavingsMonthly,
        priceDifferenceUsd: roundTo(rule.priceDifferenceUsd, 2),
        difficulty: rule.difficulty,
      });
      totalPotentialSavingsMonthly = roundTo(
        totalPotentialSavingsMonthly + co2eSavingsMonthly,
        2
      );
    }

    return {
      wallet: request.wallet,
      suggestions,
      totalPotentialSavingsMonthly,
    };
  }
}

export const suggestionsService = new SuggestionsService();

File: app/services/suggestions_service.py
"""Deterministic swap suggestions with optional narrated phrasing."""

from __future__ import annotations

from decimal import Decimal
from hashlib import sha256

from contracts.python.contracts import (
    DEFAULT_TRANSACTION_LIMIT,
    EmissionCategory,
    SwapSuggestion,
    SwapSuggestionsRequest,
    SwapSuggestionsResponse,
)

from app.core.constants import CATEGORY_ORDER, CATEGORY_ORDER_INDEX, SUGGESTION_RULES
from app.core.deterministic import decimal_to_float, round_decimal
from app.models.domain import SuggestionNarrativeInput
from app.services.emissions_service import emissions_service
from app.services.narrator_service import openai_narrator, template_narrator
from app.state.store import user_store


class SuggestionsService:
    def __init__(self) -> None:
        self._openai_narrator = openai_narrator
        self._template_narrator = template_narrator

    def _rank_categories(self, category_totals: dict[EmissionCategory, Decimal], categories: list[EmissionCategory] | None) -> list[EmissionCategory]:
        requested = categories or CATEGORY_ORDER
        ranked = sorted(
            requested,
            key=lambda category: (-category_totals[category], CATEGORY_ORDER_INDEX[category]),
        )
        if len(ranked) >= 3:
            return ranked[:5]

        seen = set(ranked)
        backfill = [
            category
            for category in sorted(
                CATEGORY_ORDER,
                key=lambda item: (-category_totals[item], CATEGORY_ORDER_INDEX[item]),
            )
            if category not in seen
        ]
        return (ranked + backfill)[:5]

    def _narrate(self, wallet: str, payload: SuggestionNarrativeInput):
        cache_key = sha256(
            (
                f"{payload.current_category.value}|{payload.current_co2e_monthly}|"
                f"{payload.alternative_co2e_monthly}|{payload.co2e_savings_monthly}|"
                f"{payload.price_difference_usd}|{payload.difficulty.value}"
            ).encode("utf-8")
        ).hexdigest()
        cached = user_store.get_narration(wallet, cache_key)
        if cached:
            return cached

        narrator = self._template_narrator
        if self._openai_narrator.is_configured():
            try:
                narrated = self._openai_narrator.narrate(payload)
            except Exception:
                narrated = narrator.narrate(payload)
        else:
            narrated = narrator.narrate(payload)
        user_store.set_narration(wallet, cache_key, narrated)
        return narrated

    def get_swap_suggestions(self, request: SwapSuggestionsRequest) -> SwapSuggestionsResponse:
        snapshot = emissions_service.get_canonical_snapshot(request.wallet)
        ranked_categories = self._rank_categories(snapshot.category_emission_totals, request.categories)

        suggestions: list[SwapSuggestion] = []
        total_savings = Decimal("0")
        for category in ranked_categories:
            rule = SUGGESTION_RULES[category]
            current_co2e = snapshot.category_emission_totals[category]
            alternative_co2e = round_decimal(current_co2e * rule.alternative_share, 2)
            savings = round_decimal(current_co2e - alternative_co2e, 2)

            narration_input = SuggestionNarrativeInput(
                current_category=category,
                current_description=rule.current_description,
                alternative_description=rule.alternative_description,
                current_co2e_monthly=current_co2e,
                alternative_co2e_monthly=alternative_co2e,
                co2e_savings_monthly=savings,
                price_difference_usd=rule.price_difference_usd,
                difficulty=rule.difficulty,
            )
            narration = self._narrate(request.wallet, narration_input)

            suggestions.append(
                SwapSuggestion(
                    current_category=category,
                    current_description=narration.current_description,
                    current_co2e_monthly=decimal_to_float(current_co2e, 2),
                    alternative_description=narration.alternative_description,
                    alternative_co2e_monthly=decimal_to_float(alternative_co2e, 2),
                    co2e_savings_monthly=decimal_to_float(savings, 2),
                    price_difference_usd=decimal_to_float(rule.price_difference_usd, 2),
                    difficulty=rule.difficulty,
                )
            )
            total_savings += savings

        return SwapSuggestionsResponse(
            wallet=request.wallet,
            suggestions=suggestions,
            total_potential_savings_monthly=decimal_to_float(total_savings, 2),
        )


suggestions_service = SuggestionsService()
